### ISSUE TITLE
Rubocop Configuration - Phase 1

### ISSUE THREAD

**Issue by Adeynack** [nerdgeschoss/shimmer#23]
# Acceptance Criteria

## Phase 1

- [x] New cops, outside of _Standard RB_ are deactivated by default
- [x] Find a way to opt-in segments of _Rubocop_ (_Rails_, _Performance_, _Rake_, _GraphQL_), so projects that do not have _GraphQL_ should not have to pull its `rubocop-graphql` for no reason.
- [x] Relevant Cops for _RSpec_ are enabled explicitly for
  - [x] `-rspec`
  - [x] `-rails`
  - [x] `-graphql`
- [ ] Following projects are referencing _Shimmer_'s rubocop and are stripped down of customized cops and have PRs open (waiting for this one) where the new styling is applied.
  - [x] KS Core ([Issue](https://github.com/KrasserStoff/core/issues/1528) / [Pull Request](https://github.com/KrasserStoff/core/pull/1545))

## Phase 2

See #42.

**Comment by Adeynack** [comment]
@JensRavens wrote [here](https://github.com/nerdgeschoss/copa-coupona/pull/272#discussion_r1026199852)

> ok, after researching the topic I agree: it's a good rule to keep. but that doesn't invalidate my former point: I don't see this rule explicitly enabled anywhere in the config but it still applies - which means we're applying rules without discussing them upfront.

@Adeynack wrote [here](https://github.com/nerdgeschoss/copa-coupona/pull/272#discussion_r1027855003)

> That was concerning styles and further extensions (RSpec, Rake, GraphQL), but my understanding from last time we discussed the Rails one was that they would stick on by default, to help us all, especially the more junior employees, to know what's good and not in Rails. We can disable them all by default (still talking about the Rails ones specifically here), but then we need to really go through the list and enable quite a lot of them. The Rails cops are usually right and we always worked with them turned on since I'm at the company, and they helped me learn lots (ex: don't forget dependent, provide inverse_of) and then we actually just learned about a new one with that BulkChangeTable. I see it as an amazing learning tool and that's why I would advocate to leave Rails cops on by default. For the rest, I am totally OK with turning them off as we of course agreed to.

Let's discuss this here then, @JensRavens :-) 

And input from other Rails developers @oma-s @Maxi211 @jmarsh24 is appreciated of course, since it affects how the whole team works.

**Comment by Adeynack** [comment]
# GraphQL

```yaml
GraphQL:
  Enabled: false

GraphQL/OrderedFields:
  Enabled: true
  Exclude:
    - "*"
  Include:
    - '**/mutation_type.rb'

GraphQL/ArgumentName:
  Enabled: true

GraphQL/ArgumentUniqueness:
  Enabled: true

GraphQL/FieldHashKey:
  Enabled: true

GraphQL/FieldMethod:
  Enabled: true

GraphQL/FieldName:
  Enabled: true

GraphQL/FieldUniqueness:
  Enabled: true

GraphQL/LegacyDsl:
  Enabled: true

GraphQL/MultipleFieldDefinitions:
  Enabled: true

GraphQL/NotAuthorizedNodeType:
  Enabled: true

GraphQL/UnnecessaryArgumentCamelize:
  Enabled: true

GraphQL/UnnecessaryFieldAlias:
  Enabled: true

GraphQL/UnnecessaryFieldCamelize:
  Enabled: true

GraphQL/UnusedArgument:
  Enabled: true
```

**Comment by Adeynack** [comment]
@JensRavens @oma-s @Maxi211 @jmarsh24

Dear Back-Enders, here's a suggestion of ruleset to consider our own. It contains for the moment:

- the basics
- `rspec`
- `rails`
- `graphql`
- a mechanism for not activating rules unless their GEM is detected

That ruleset is applied for the moment in [this pull request on Krasserstoff](https://github.com/KrasserStoff/core/pull/1545).

Look at the description of this issue to see how I propose we proceed.

**Comment by Adeynack** [comment]
I tagged a few of you (back-end-ers) to this issue, but I somehow feel it needs a bit more of a “live” (well, chat that is) introduction.
https://github.com/nerdgeschoss/shimmer/issues/23
The idea of this first pull request is that we all debate — if needed — every rule I suggest to add on top of standard. It’s a rough lot of them, but we need to start somewhere. And here’s your chance to debate if you want a rule or not, if you want it only as a warning but not as a CI-failing rule (it’s possible with rubocop), if you want it but configured differently… and when we agree on them, they will be applied to all project we have with Ruby code. So here’s your chance to have a saying in how our code will be lint across the company.

And to be clear, even though there are quite a few that I am proposing here (that I do believe are helpful), the core approach remains:
we go with that Standard RB is proposing (no bike-shedding there)
we add,  with a PR to debate it, manually the rules that we believe will help us work on our codebases

So if you do feel any of those I am proposing are slightly too much, then by all mean, we can debate and remove them and keep things more flexible.
The goal here is that Rubocop…
helps us keep our code clean
teaches us a few things along the way (I found it to be a good “teacher” on many things Rails)
remind us of some potential basic mistakes we could be doing

So go, read, and start threads on those rules so we can debate and agree on them! :slightly_smiling_face: 
I hope this makes things clearer.
